context/AppDataContext.tsx

```diff
@@ -62,13 +69,38 @@ const AppDataProvider = ({ children }: { children: React.ReactNode }) => {
     setRepositories(updatedRepositories);
   };
 
+  const filterRepositoriesByTag = (tag: string) => {
+    return repositories.filter((repository) =>
+      repository.tags?.some((t) => t.id === tag)
+    );
+  };
+
+  // Define the filter function for the query
+  const filterRepositoriesByQuery = (query: string) => {
+    if (!query) {
```

Comment:
I would make sure there are at least 3 chars in the search box before calling this method instead. Searching like we're doing is expensive and it's best if we do it as little as possible.